### ISSUE TITLE
Includes not working in AspNetCore 2.1.1 App when hosted as a service

### ISSUE THREAD

**Issue by Jacrys** [aspnet/Docs#7445]
I have an app which is being hosted as a service which was built as so:
`sc create DashboardService binPath= "c:\repos\dashboard\bin\Release\netcoreapp2.1\win10-x64\publish\dashboard.exe"`
when I navigate to the endpoint, however, the page renders incorrectly and I get the below errors:
`Loading failed for the &lt;script&gt; with source “http://localhost:5000/lib/signalr/signalr.js”.
Loading failed for the &lt;script&gt; with source “http://localhost:5000/js/chat.js”.
Loading failed for the &lt;script&gt; with source “http://localhost:5000/js/site.min.js”.`
which leads me to believe none of the includes are working correctly (this is backed up by the CSS not working correctly).  

Note: I get 404 errors for all of the assets that are loaded locally, but the publish did include them in the publish folder under the wwwroot folder as expected.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: b0bd8cb8-7e28-8c4d-a550-3aea8f3ab4dd
* Version Independent ID: 4b60d24e-ef05-15e5-10b4-483094081bf3
* Content: [Host ASP.NET Core in a Windows Service](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/windows-service?view=aspnetcore-2.1)
* Content Source: [aspnetcore/host-and-deploy/windows-service.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/host-and-deploy/windows-service.md)
* Product: **aspnet-core**
* GitHub Login: @guardrex
* Microsoft Alias: **tdykstra**

**Comment by Jacrys** [comment]
Yes, Static File Middleware is enabled.

`using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.HttpsPolicy;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;

namespace dashboard
{
    public class Startup
    {
        public Startup(IConfiguration configuration)
        {
            Configuration = configuration;
        }

        public IConfiguration Configuration { get; }

        // This method gets called by the runtime. Use this method to add services to the container.
        public void ConfigureServices(IServiceCollection services)
        {
            services.Configure<CookiePolicyOptions>(options =>
            {
                // This lambda determines whether user consent for non-essential cookies is needed for a given request.
                options.CheckConsentNeeded = context => true;
                options.MinimumSameSitePolicy = SameSiteMode.None;
            });

            services.AddSignalR();
            services.AddMvc().SetCompatibilityVersion(CompatibilityVersion.Version_2_1);
        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IHostingEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }
            else
            {
                app.UseExceptionHandler("/Home/Error");
                app.UseHsts();
            }

            app.UseHttpsRedirection();
            app.UseStaticFiles();
            app.UseCookiePolicy();
            app.UseSignalR(routes =>
            {
                routes.MapHub<ChatHub>("/chatHub");
            });

            app.UseMvc(routes =>
            {
                routes.MapRoute(
                    name: "default",
                    template: "{controller=Home}/{action=Index}/{id?}");
            });
        }
    }
}
`
Do I need to specifically add the static files assembly to using?  The static files docs make it sound fairly implicit when targeting .NET Core.

**Comment by guardrex** [comment]
> Do I need to specifically add the static files assembly to using?

No worries there ... it's available via `Microsoft.AspNetCore.Builder`.

You're sure it's a 404, right? What happens if you hit it at `https://localhost:5000/js/site.min.js` (the secure endpoint)?

**Comment by Jacrys** [comment]
> You're sure it's a 404, right? 
![static404s](https://user-images.githubusercontent.com/3784287/42348320-a48c8732-806e-11e8-98e2-601956a3e18d.PNG)

> What happens if you hit it at https://localhost:5000/js/site.min.js (the secure endpoint)?
![css404](https://user-images.githubusercontent.com/3784287/42348332-af859dd6-806e-11e8-81c2-e423286d22ce.PNG)

**Comment by guardrex** [comment]
I was curious to know if the HTTPS Redirection Middleware had anything to do with it.

@Tratcher have you seen this behavior before? We didn't hit it with the 2.1 WS samp.

**Comment by Tratcher** [comment]
Sounds like the ContentRoot is wrong. The default GetCurrentDirectory doesn't work for Windows Services. How are you setting it?

**Comment by guardrex** [comment]
Just to clarify @Tratcher, you're asking about something in relation to `GetCurrentDirectory` that doesn't have a bearing on our sample app, correct? I can't repro @Jacrys behavior. When I run the WS sample app as a service, I'm able to get the static assets to load ...

```html
<link rel="stylesheet" href="/css/site.min.css?v=DlTdkceTKP_cDKN2ooajjMLy5VAa_6B26xGDSW6LE1k">
```

<img width="346" alt="capture" src="https://user-images.githubusercontent.com/1622880/42351895-ed7c9d02-807c-11e8-9b3f-5169db4cd02a.PNG">

`http://localhost:5000/css/site.min.css` correctly yields ...

```
body{padding-top:50px;padding-bottom:20px}h1{font-size:30px}h2{font-size:24px}.body-content{padding-left:15px;padding-right:15px}.panel-body{font-size:16px}
```

**Comment by Tratcher** [comment]
Our sample wouldn't be able to use GetCurrentDirectory either, the instructions say not to:
https://github.com/aspnet/Docs/blob/master/aspnetcore/host-and-deploy/windows-service.md#get-started

I'm just wondering how well he followed the instructions.

**Comment by Jacrys** [comment]
I am not explicitly setting it, but I guess maybe parameterless UseStaticFiles uses a Directory.GetCurrentDirectory() as the default option?  That is merely speculation on my part as digging through the source has yielded little.  I do see it inherits from SharedOptionsBase and instantiates a new SharedOptions object, but the default constructor for that appears to set requestPath to an empty PathString.  I therefore changed Startup.cs to explicitly reference the publish folder:

    app.UseStaticFiles(requestPath: AppDomain.CurrentDomain.BaseDirectory) 

That iteration would not even start up. I then changed it to reference the wwwroot folder thinking that might be the issue, but the same result, the service would not start.

**Comment by guardrex** [comment]
@Jacrys Did you pull down the topic's sample app to confirm it working in the general sense? If you can get `http://localhost:5000/css/site.min.css` from the sample app, then there's definitely some kind of delta between the sample app and your app causing this behavior.

**Comment by Tratcher** [comment]
See UseContentRoot as described here: https://github.com/aspnet/Docs/blob/master/aspnetcore/host-and-deploy/windows-service.md#get-started
Also, try a hardcoded path to your application root (not your to your static files) to make sure that's right.

@guardrex what's up with the `::: moniker range=">= aspnetcore-2.0"`?

**Comment by guardrex** [comment]
2.x ...

```csharp
public static void Main(string[] args)
{
    var pathToExe = Process.GetCurrentProcess().MainModule.FileName;
    var pathToContentRoot = Path.GetDirectoryName(pathToExe);

    var host = WebHost.CreateDefaultBuilder(args)
        .UseContentRoot(pathToContentRoot)
        .UseStartup<Startup>()
        .Build();

    host.RunAsService();
}
```

vs 1.x ...

```csharp
public static void Main(string[] args)
{
    var pathToExe = Process.GetCurrentProcess().MainModule.FileName;
    var pathToContentRoot = Path.GetDirectoryName(pathToExe);

    var host = new WebHostBuilder()
        .UseKestrel()
        .UseContentRoot(pathToContentRoot)
        .UseIISIntegration()
        .UseStartup<Startup>()
        .Build();

    host.RunAsService();
}
```

... we go with `WebHostBuilder` for 1.x ..... original topic is 3/30/2017 ... so it's 1.x viable, correct?

**Comment by Jacrys** [comment]
The kicker (and why I ignored that line) is, that the web service I was running as a Windows service did not call UseContentRoot.  The doc specifically states 
> If the code calls `UseContentRoot`, use a path to the app's published location instead of `Directory.GetCurrentDirectory()`.

but includes no contingency if it does not call `UseContentRoot`.  So I inferred that since that section did not apply to my situation, I did not need to follow the steps outlined there.  This is what the WebHostBuilder looked like:

    WebHost.CreateDefaultBuilder(args).UseStartup<Startup>()

After adding `UseContentRoot` method as above (and in doc) it works!  

It probably would be good to clarify that point in the documentation.  (The example I was loading into a service was the example from [Create a real-time SignalR web app](https://docs.microsoft.com/en-us/aspnet/core/tutorials/signalr?view=aspnetcore-2.1&tabs=visual-studio-code,) which uses the `dotnet new webapp` command to initialize the files.  It apparently does not explicitly call `UseContentRoot()`.
